File: apps/server/agent/src/memory/registry.ts
/**
 * Resolves (scope, projectId) to the right SqliteMemoryStorage instance.
 * `tools/memory.ts` only ever talks to this module — never opens a DB file
 * directly.
 *
 * Layout:
 *   <storage>/projects/<projectId>/memory.db   (scope: "project")
 *   <storage>/memory-global.db                 (scope: "global", one shared file)
 */
import path from "node:path";
import { getConsoleStorageDir } from "../session/apppaths.js";
import { SqliteMemoryStorage } from "./storage.js";

const MAX_CACHED_PROJECT_STORES = 50;

export function getProjectMemoryDbPath(storageDir: string, projectId: string): string {
  return path.join(storageDir, "projects", projectId, "memory.db");
}

export function getGlobalMemoryDbPath(storageDir: string): string {
  return path.join(storageDir, "memory-global.db");
}

export class MemoryRegistry {
  private readonly storageDir: string;
  private readonly projectStores = new Map<string, SqliteMemoryStorage>();
  private globalStore: SqliteMemoryStorage | null = null;

  constructor(storageDir: string = getConsoleStorageDir()) {
    this.storageDir = storageDir;
  }

  forProject(projectId: string): SqliteMemoryStorage {
    const cached = this.projectStores.get(projectId);
    if (cached) {
      this.projectStores.delete(projectId);
      this.projectStores.set(projectId, cached);
      return cached;
    }
    const dbPath =
      this.storageDir === ":memory:" ? ":memory:" : getProjectMemoryDbPath(this.storageDir, projectId);
    const store = new SqliteMemoryStorage(dbPath, "project");
    this.projectStores.set(projectId, store);
    if (this.projectStores.size > MAX_CACHED_PROJECT_STORES) {
      const oldest = this.projectStores.keys().next().value as string | undefined;
      if (oldest !== undefined && oldest !== projectId) {
        this.projectStores.get(oldest)?.close();
        this.projectStores.delete(oldest);
      }
    }
    return store;
  }

  forGlobal(): SqliteMemoryStorage {
    if (!this.globalStore) {
      const dbPath = this.storageDir === ":memory:" ? ":memory:" : getGlobalMemoryDbPath(this.storageDir);
      this.globalStore = new SqliteMemoryStorage(dbPath, "global");
    }
    return this.globalStore;
  }

  resolve(scope: "project" | "global", projectId: string | null): SqliteMemoryStorage {
    if (scope === "global") return this.forGlobal();
    if (!projectId) {
      throw new Error("Memory scope 'project' requires a projectId; this session has no project.");
    }
    return this.forProject(projectId);
  }
}

File: apps/server-go/internal/agent/memory/registry.go
// Memory store registry. Port of agent/src/memory/registry.ts: resolves
// (scope, projectId) to a cached store instance with LRU eviction.
package memory

import (
	"container/list"
	"path/filepath"
	"sync"

	"github.com/Adelodunpeter25/console/apps/server-go/internal/utils"
)

// MaxCachedProjectStores mirrors the TS LRU cap.
const MaxCachedProjectStores = 50

// Registry opens memory databases on demand:
// <storage>/projects/<projectId>/memory.db (project scope) and
// <storage>/memory-global.db (global scope).
type Registry struct {
	mu         sync.Mutex
	storageDir string
	global     *Store
	projects   map[string]*list.Element
	lru        *list.List
}

type projectEntry struct {
	id    string
	store *Store
}

// NewRegistry creates a registry rooted at dir ("" resolves the console
// storage dir, mirroring getConsoleStorageDir).
func NewRegistry(dir string) *Registry {
	if dir == "" {
		dir = utils.ConsoleStorageDir()
	}
	return &Registry{storageDir: dir, projects: map[string]*list.Element{}, lru: list.New()}
}

// ProjectMemoryDBPath returns the project scope database path.
func ProjectMemoryDBPath(storageDir, projectID string) string {
	return filepath.Join(storageDir, "projects", projectID, "memory.db")
}

// GlobalMemoryDBPath returns the shared global database path.
func GlobalMemoryDBPath(storageDir string) string {
	return filepath.Join(storageDir, "memory-global.db")
}

// StorageDir returns the registry root (tests).
func (r *Registry) StorageDir() string { return r.storageDir }

// ForProject returns the cached (or newly opened) project store.
func (r *Registry) ForProject(projectID string) (*Store, error) {
	r.mu.Lock()
	defer r.mu.Unlock()
	if el, ok := r.projects[projectID]; ok {
		r.lru.MoveToFront(el)
		return el.Value.(*projectEntry).store, nil
	}
	store, err := OpenStore(ProjectMemoryDBPath(r.storageDir, projectID), ScopeProject)
	if err != nil {
		return nil, err
	}
	r.projects[projectID] = r.lru.PushFront(&projectEntry{id: projectID, store: store})
	for r.lru.Len() > MaxCachedProjectStores {
		back := r.lru.Back()
		if back == nil {
			break
		}
		victim := back.Value.(*projectEntry)
		if victim.id == projectID {
			break
		}
		victim.store.Close()
		delete(r.projects, victim.id)
		r.lru.Remove(back)
	}
	return store, nil
}

// ForGlobal returns the shared global store.
func (r *Registry) ForGlobal() (*Store, error) {
	r.mu.Lock()
	defer r.mu.Unlock()
	if r.global != nil {
		return r.global, nil
	}
	store, err := OpenStore(GlobalMemoryDBPath(r.storageDir), ScopeGlobal)
	if err != nil {
		return nil, err
	}
	r.global = store
	return store, nil
}

// Resolve returns the store for a scope, requiring a project id for the
// project scope (mirrors the TS resolve error).
func (r *Registry) Resolve(scope Scope, projectID string) (*Store, error) {
	if scope == ScopeGlobal {
		return r.ForGlobal()
	}
	if projectID == "" {
		return nil, errNoProject()
	}
	return r.ForProject(projectID)
}

type noProjectError struct{}

func (e *noProjectError) Error() string {
	return "Memory scope 'project' requires a projectId; this session has no project."
}

func errNoProject() error { return &noProjectError{} }
